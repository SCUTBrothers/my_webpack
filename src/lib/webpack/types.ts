export interface ResolveOptions {
  /**
   * @description
   * 用于指定额外的搜索node_modules的路径范围
   */
  paths?: string[]
  extensions?: string[]
}

export interface Options {
  resolve?: ResolveOptions
  // JSONP function used to load chunks
  outputJsonpFunction?: string
  // Path from where chunks are loaded
  scriptSrcPrefix?: string
  // write files to this directory (absolute path)
  outputDirectory?: string
  // write first chunk to this file
  output?: string
  // write chunks to files named chunkId plus outputPostfix
  outputPostfix?: string
  // exports of input file are stored in this variable
  library?: string
  // minimize outputs with uglify-js
  minimize?: boolean
  // add absolute filenames of input files as comments
  includeFilenames?: boolean
}

export interface BufferStat {
  chunkCount: number
  modulesCount: number
  modulesIncludingDuplicates: number
  modulesPerChunk: number
  modulesFirstChunk: number
  fileSizes: Record<string, number>
}

export type OutPut = string | BufferStat
