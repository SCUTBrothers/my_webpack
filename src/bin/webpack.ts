#!/usr/bin/env node

import path from 'path'
import fs from 'fs'
import meow from 'meow/build'
import webpack from '../lib/webpack'
import { BufferStat, Options } from '../lib/webpack/types'

const cli = meow(
  `
  Usage: $0 <input> <output>
  
  Options:
    --single, -s  Disable Code Splitting (default: false)
    --min, -m     Minimize it with uglifyjs (default: false)
    --filenames   Output Filenames Into File (default: false)
    --options     Options JSON      
    --script-src-prefix  Path Prefix For JavaScript Loading
    --library     Stores the exports into this variable  
`,
  {
    importMeta: import.meta,
    flags: {
      single: {
        type: 'boolean',
        alias: 's'
      },
      min: {
        type: 'boolean',
        alias: 'm'
      },
      filenames: {
        type: 'boolean'
      },
      options: {
        type: 'string'
      },
      'script-src-prefix': {
        type: 'string'
      },
      library: {
        type: 'string'
      }
    }
  }
)

let [input, output] = cli.input

if (input && input[0] !== '/' && input[1] !== ':') {
  input = path.join(process.cwd(), input)
}
if (output && output[0] !== '/' && input[1] !== ':') {
  output = path.join(process.cwd(), output)
}

const { flags } = cli

let options: Options = {}

if (flags.options) {
  options = JSON.parse(fs.readFileSync(flags.options, 'utf-8')) as Record<string, unknown>
}

if (flags['script-src-prefix']) {
  options.scriptSrcPrefix = cli.flags['script-src-prefix'] as string | undefined
}

if (flags.min) {
  options.minimize = true
}

if (flags.filenames) {
  options.includeFilenames = true
}

if (flags.libary) {
  options.library = flags.library
}

if (flags.single) {
  webpack(input, options, (err, source) => {
    if (err) {
      console.error(err)
      return
    }
    if (output) {
      fs.writeFileSync(output, source as string, 'utf-8')
    } else {
      process.stdout.write(source as string)
    }
  })
} else {
  output = output || path.join(process.cwd(), 'js', 'web.js')
  if (!options.outputDirectory) options.outputDirectory = path.dirname(output)
  if (!options.output) options.output = path.basename(output)
  if (!options.outputPostfix) options.outputPostfix = `.${path.basename(output)}`

  const outExists = fs.existsSync(options.outputDirectory)

  if (!outExists) fs.mkdirSync(options.outputDirectory)

  webpack(input, options, (err, stats) => {
    if (err) {
      console.error(err)
      return
    }
    console.log(stats as BufferStat)
  })
}
