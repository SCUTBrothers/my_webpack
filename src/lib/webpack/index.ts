import path from 'path'
import fs from 'fs'
// import uglify from 'uglify-js'
import { BufferStat, Options, OutPut } from './types'
import { Callback } from '../utils/types'
import buildDeps from '../buildDeps'
import writeChunk from '../write'

const templateAsync = fs.readFileSync(path.join(__dirname, 'templateAsync.js')).toString()
const templateSingle = fs.readFileSync(path.join(__dirname, 'templateSingle.js')).toString()

function uglifyFunc(input: string, filename: string) {
  console.log('\n üéØ-> Line 22 check the variable filename: üìÆÔ∏è---- üìÆÔ∏è', filename)
  // try {
  //   source = uglify.parser.parse(input)
  //   source = uglify.uglify.ast_mangle(source)
  //   source = uglify.uglify.ast_squeeze(source)
  //   source = uglify.uglify.gen_code(source)
  // } catch (e) {
  //   console.error(`${filename} @ Line ${e.line}, Col ${e.col}, ${e.message}`)
  //   return input
  // }
  //
  // return source
  return input
}

function stringify(str: string) {
  return `"${str.replace(/\\/g, '\\\\').replace(/"/g, '\\"')}"`
}

/*

	callback: function(err, source / stats)
	  source if options.output is not set
	  else stats json
*/
export default function webpack(
  context: string,
  moduleName: string,
  options: Options,
  callback: Callback<OutPut>
): void
export default function webpack(
  context: string,
  moduleName: string,
  callback: Callback<OutPut>
): void
export default function webpack(
  contextOrAbsoluteModulePath: string,
  options: Options,
  callback: Callback<OutPut>
): void
export default function webpack(
  contextOrAbsoluteModulePath: string,
  callback: Callback<OutPut>
): void
export default function webpack(
  contextOrAbsoluteModulePath: string,
  moduleNameOrOptionsOrCallback: string | Options | Callback<OutPut>,
  optionsOrCallback?: Options | Callback<OutPut>,
  maybeEmptyCallback?: Callback<OutPut>
) {
  let context: string
  let moduleName: string
  let options: Options
  let callback: Callback<OutPut>
  if (!optionsOrCallback && !maybeEmptyCallback) {
    // 2 args
    context = path.dirname(contextOrAbsoluteModulePath)
    moduleName = `./${path.basename(contextOrAbsoluteModulePath)}`
    options = {} as Options
    callback = moduleNameOrOptionsOrCallback as Callback<OutPut>
  } else if (!maybeEmptyCallback) {
    // 3 args
    if (typeof moduleNameOrOptionsOrCallback === 'string') {
      context = contextOrAbsoluteModulePath
      moduleName = moduleNameOrOptionsOrCallback
      options = {} as Options
    } else {
      context = path.dirname(contextOrAbsoluteModulePath)
      moduleName = `./${path.basename(contextOrAbsoluteModulePath)}`
      options = moduleNameOrOptionsOrCallback as Options
    }
  } else {
    // 4 args
    context = contextOrAbsoluteModulePath
    moduleName = moduleNameOrOptionsOrCallback as string
    options = optionsOrCallback as Options
    callback = maybeEmptyCallback
  }

  buildDeps(context, moduleName, options, (err, depTree) => {
    if (err) {
      callback(err, null)
      return
    }

    if (!depTree) {
      callback(new Error('No depTree'), null)
      return
    }

    let buffer: string = ''
    const fileSizeMap: Record<string, number> = {}

    if (options.output) {
      const defaultOptions = {
        outputJsonpFunction: `webpackJsonp${options.library || ''}`,
        scriptSrcPrefix: '',
        outputDirectory: path.dirname(options.output),
        output: path.basename(options.output),
        outputPostfix: `.${options.output}`
      }

      options = { ...defaultOptions, ...options }

      let chunksCount = 0
      const chunkIds = Object.keys(depTree.chunks)
      chunkIds.forEach((chunkId) => {
        const chunk = depTree.chunks[chunkId]
        if (chunk.empty) return
        if (chunk.equals !== undefined) return

        chunksCount++
        const filename = path.join(
          options.outputDirectory!,
          chunk.id === 0 ? options.output! : chunk.id + options.outputPostfix!
        )

        if (chunk.id === 0) {
          if (options.library) {
            buffer += `var ${options.library}=`
          }
          if (Object.keys(depTree.chunks).length > 1) {
            buffer += `
            ${templateAsync}
            /******/({a: ${stringify(options.outputPostfix!)},
                      b: ${stringify(options.outputJsonpFunction!)},
                      c: ${stringify(options.scriptSrcPrefix!)},
            `
          } else {
            buffer += `${templateSingle}
              /******/({`
          }
        } else {
          buffer += `/******/${options.outputJsonpFunction!}(${chunk.id}, {`
        }
        buffer += writeChunk(depTree, chunk, options)
        buffer += '/******/})'
        if (options.minimize) buffer = uglifyFunc(buffer, filename)

        fs.writeFile(filename, buffer, 'utf-8', (writeFileError) => {
          if (writeFileError) throw writeFileError
        })
        fileSizeMap[path.basename(filename)] = buffer.length
      })

      const modulesIncludingDuplicates = Object.keys(depTree.chunks).reduce((sum, chunkId) => {
        const modulesIds = Object.keys(depTree.chunks[chunkId].modules)
        modulesIds.forEach((moduleId) => {
          if (depTree.chunks[chunkId].modules[moduleId] === 'include') sum++
        })
        return sum
      }, 0)

      const bufferStat: BufferStat = {
        chunkCount: chunksCount,
        modulesCount: Object.keys(depTree.modulesById).length,
        modulesIncludingDuplicates,
        modulesPerChunk: Math.round((modulesIncludingDuplicates / chunksCount) * 10) / 10,
        modulesFirstChunk: Object.keys(depTree.chunks[0].modules).reduce((sum, moduleId) => {
          if (depTree.chunks[0].modules[moduleId] === 'include') sum++
          return sum
        }, 0),
        fileSizes: fileSizeMap
      }
      callback(null, bufferStat)
    } else {
      buffer = `
      ${options.library ? `var ${options.library}=` : ''}
      ${templateSingle}
      /******/({
      ${writeChunk(depTree, options)}
      /******/})
      `
      if (options.minimize) buffer = uglifyFunc(buffer, 'output')
      callback(null, buffer)
    }
  })
}
