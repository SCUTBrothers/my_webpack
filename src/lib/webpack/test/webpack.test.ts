import path from 'path'
import webpack from '../index'

const fixtures = path.resolve(__dirname, 'fixtures')

describe('webpack', () => {
  it('should return a single bundle which contains all sync modules', (done) => {
    webpack(fixtures, './simple/module1', (err, output) => {
      expect(err).toBeNull()
      expect(output).toMatchSnapshot()
      console.log('\n 🎯-> Line 11 check the variable output: 📮️---- 📮️', output)
      done()
    })
  })
})
