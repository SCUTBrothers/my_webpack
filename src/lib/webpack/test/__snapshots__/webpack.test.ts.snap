// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`webpack should return a single bundle which contains all sync modules 1`] = `
"
      
      /******/ ;(function (document, undefined) {
  /******/ return function (modules) {
    /******/ var installedModules = {}
    /******/ function require(moduleId) {
      /******/ if (installedModules[moduleId]) /******/ return installedModules[moduleId]
      /******/ var module = (installedModules[moduleId] = {
        /******/ exports: {}
        /******/
      })
      /******/ modules[moduleId](module, module.exports, require)
      /******/ return module.exports
      /******/
    }
    /******/ require.ensure = function (chunkId, callback) {
      /******/ callback(require)
      /******/
    }
    /******/ return require(0)
    /******/
  }
  /******/
})(document)

      /******/({
      
    /******/
    0: function(module, exports, require) {
    
    var a = require(1)
if (x) {
  for (var i = 0; i < 100; i++) {
    while (true) require(2)
    do {
      i++
    } while (require(3)())
  }
}

    /******/},
    
    /******/
    1: function(module, exports, require) {
    
    module.exports = function a() {
  return 'This is a'
}

    /******/},
    
    /******/
    2: function(module, exports, require) {
    
    module.exports = function b() {
  return 'This is b'
}

    /******/},
    
    /******/
    3: function(module, exports, require) {
    
    module.exports = function a() {
    return "This is a in m1";
};

    /******/},
    
      /******/})
      "
`;
